### ISSUE TITLE
Question: Process Concurrency in v4

### ISSUE THREAD

**Issue by kuiperzone** [litedb-org/LiteDB#1076]
But first, let me say well done on v4. It's looking great!

I would like to have one process which writes to an LiteDB file, and one or two which read from it. The writer process will be inserting around one document per minute, whereas the reader(s) which will run occasionally but will want to read through a subset of documents as fast as possible.

1. In the wiki, it says the file is locked when it is read/or written to. What happens when a call trys to read or write when it is locked? Does it throw an exception, does it wait? or does it cache data to be flushed to file later?

2. What setting string should I use to best support my scenario?

Thanks

**Comment by mbdavid** [comment]
Hi @kuiperzone,

If you have 2 different process (not threads) you need open 2 different instances. In each process, you can keep database instance always open so you can use cached pages in memory.

About locks, there is no exceptions - just need wait finish write process to start read process. You will got exception only if you need wait more than 1 minute (of Timeout definition)

In v4, you can have multiples readers at same time, but only 1 writer (with no readers) per time.. all this "per database". There is no transaction in v4, so this write time must be as short as possible.

In v5, thanks for WAL, you have more possibilities, like write locks per collections an no lock in read (even if another thread are writing)

**Comment by kuiperzone** [comment]
Thanks David.

**Comment by PaluMacil** [comment]
> When will be the v5 is coming?

Currently the README in the v5 branch says late 2018.

**Comment by mbdavid** [comment]
I will release first beta version in about one month.... current v5 version are 95% defined and developed... but i'm still fixing some bugs, adding more tests... and the hardest part: update all documentation 😄

**Comment by kuiperzone** [comment]
What are the major changes in v5?

**Comment by mbdavid** [comment]
v5 has a lot of changes and that why I working almost a year.

- New WAL (write ahead log) for better ACID implementation (and much more fast write)
- Support isolated transactions
- Write locks per collection - no locks for readers
- Support data version (transactions are contains initial data version)
- New `BsonExpressions` if support in all predicates
- New SQL language for insert/update/delete/indexes/.... and queries
- New system $collections (some like INFORMATION_SCHEMA)
- Support `OrderBy` on engine
- Support aggregation based on expressions (`GroupBy` syntax) [with min/max/count/avg/first/last]
- Complete new Linq To `BsonExpression` for better use of Linq (with support to methods)
- New virtual input/output function collection ("get-from" or "write-to" data from SQL to json/text/binary file)
- BSON partial deserialized fields (read/deserialize only used fields)
- New LiteDB.Studio - WinForms UI for manage with multi-thread (with tabs) support.
- ....

Here some cool SQL queries that is already possible do in v5:

```SQL
-- SELECT with where/orderBy
SELECT $
  FROM customers
 WHERE name LIKE 'John%'
 ORDER BY lastModified DESC

-- Using INCLUDE (DbRef) 
 SELECT { _id, orderDate, name: customer.name }
   FROM orders
INCLUDE customer
  WHERE customer.age > 65
  ORDER BY lastModified DESC
 
-- GROUP BY support writing on disk
SELECT { state, count: COUNT($) }
  INTO $file_json('c:/temp/output.json')
  FROM person
 WHERE state IN ["FL", "CA", "MN", "TX", "MS"]
 GROUP BY state
HAVING count > 5;

-- UPDATE (merge) fields with functions
UPDATE customer
   SET {
           name = UPPER(name), 
           active = DATEDIFF('d', lastModified, DATE()) < 30
       }
 WHERE _id > 0;
```

There are in plans (but not for first v5 release) support for server mode (run LiteDB as service using namedpipes/http/rest/... wires protocol).

But to me, SQL + WAL are biggest changes in v5!

**Comment by mbdavid** [comment]
Hi @AwsomeCode, SQL syntax are quite simple... as simple as Shell syntax. BsonExpressions was much more complex.... and both are done. All features that I list above already done 😄 

SQL are awesome in Studio GUI tool... you can select, transform, group, order... with just a single string command. There is no "build a small code to update a collection", or check some data conditional...

**Comment by petertiedemann** [comment]
@mbdavid If i write a 100MB file to FileStorage, will it hold the lock throughout the write, or will readers get a chance to read between each 1MB chunk?

**Comment by JKamsker** [comment]
v4 is no longer developed, so here is the answer for v5 (5.0.21 and the 6.x prerelease behave the same).

- Within one process, use a single `LiteDatabase` instance. It is thread-safe, and readers don't wait for writers: they read a snapshot while the write goes to the `-log.db` file.
- For several processes on the same file, open it in shared mode in every process:

```csharp
using var db = new LiteDatabase(@"Filename=C:\data\app.db;Connection=shared");
```

In shared mode every call takes a system-wide named mutex, opens the file, runs, and closes the file again. Calls from other processes wait for the mutex. They don't throw, and nothing is cached to be flushed later. A query holds the mutex until its result has been read and disposed, so materialize it (`.ToList()`) rather than keeping an enumerator open. Because the file is reopened for each call, the page cache isn't reused between calls.

On petertiedemann's FileStorage question: in v5, `Upload` writes 255 KB chunks, and each chunk is its own insert, so other processes can get in between chunks. For a new file, the `_files` entry is written only after the last chunk, so readers don't see the file before it's complete.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
